src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/RSAPssX509SignatureGenerator.cs

```diff
@@ -0,0 +1,126 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using Internal.Cryptography;
+
+namespace System.Security.Cryptography.X509Certificates
+{
+    internal sealed class RSAPssX509SignatureGenerator : X509SignatureGenerator
+    {
+        private readonly RSA _key;
+        private readonly RSASignaturePadding _padding;
+
+        internal RSAPssX509SignatureGenerator(RSA key, RSASignaturePadding padding)
+        {
+            Debug.Assert(key != null);
+            Debug.Assert(padding != null);
+            Debug.Assert(padding.Mode == RSASignaturePaddingMode.Pss);
+
+            // Currently we don't accept options in PSS mode, but we could, so store the padding here.
+            _key = key;
+            _padding = padding;
+        }
+
+        public override byte[] GetSignatureAlgorithmIdentifier(HashAlgorithmName hashAlgorithm)
+        {
+            // If we ever support options in PSS (like MGF-2, if such an MGF is ever invented)
+            // Or, more reasonably, supporting a custom value for the salt size.
+            if (_padding != RSASignaturePadding.Pss)
+            {
+                throw new CryptographicException(SR.Cryptography_InvalidPaddingMode);
+            }
+
+            uint cbSalt;
+            string digestOid;
+
+            if (hashAlgorithm == HashAlgorithmName.SHA256)
+            {
+                cbSalt = 256 / 8;
+                digestOid = Oids.Sha256;
+            }
+            else if (hashAlgorithm == HashAlgorithmName.SHA384)
+            {
+                cbSalt = 384 / 8;
+                digestOid = Oids.Sha384;
+            }
+            else if (hashAlgorithm == HashAlgorithmName.SHA512)
+            {
+                cbSalt = 512 / 8;
+                digestOid = Oids.Sha512;
+            }
+            else
+            {
+                throw new ArgumentOutOfRangeException(
```

Comment:
This seems like a funny exception to use for "we don't know how to handle your hash algorithm in PSS"